opensearch_py_ml/ml_models/crossencodermodel.py

```diff
@@ -0,0 +1,319 @@
+        :type overwrite: bool
+        :return: None
+        """
+        default_folder_path = Path(f"/tmp/models/{hf_model_id}")
+
+        if folder_path is None:
+            self._folder_path = default_folder_path
+        else:
+            self._folder_path = Path(folder_path)
+
+        if self._folder_path.exists() and not overwrite:
+            raise Exception(
+                f"Folder {self._folder_path} already exists. To overwrite it, set `overwrite=True`."
+            )
+
+        self._hf_model_id = hf_model_id
+        self._framework = None
+        self._folder_path.mkdir(parents=True, exist_ok=True)
+        self._model_zip = None
+        self._model_config = None
+
+    def zip_model(self, framework: str = "pt", zip_fname: str = "model.zip") -> Path:
+        """
+        Compiles and zips the model to {self._folder_path}/{zip_fname}
+
+        :param framework: one of "pt", "onnx". The framework to zip the model as.
+            default: "pt"
+        :type framework: str
+        :param zip_fname: path to place resulting zip file inside of self._folder_path.
+            Example: if folder_path is "/tmp/models" and zip_path is "zipped_up.zip" then
+            the file can be found at "/tmp/models/zipped_up.zip"
+            Default: "model.zip"
+        :type zip_fname: str
+        :return: the path with the zipped model
+        :rtype: Path
+        """
+        tk = AutoTokenizer.from_pretrained(self._hf_model_id)
+        model = AutoModelForSequenceClassification.from_pretrained(self._hf_model_id)
+        features = tk([["dummy sentence 1", "dummy sentence 2"]], return_tensors="pt")
+        mname = Path(self._hf_model_id).name
+
+        # bge models don't generate token type ids
+        if mname.startswith("bge"):
+            features["token_type_ids"] = torch.zeros_like(features["input_ids"])
+
+        if framework == "pt":
+            self._framework = "pt"
+            model_loc = CrossEncoderModel._trace_pytorch(model, features, mname)
+        elif framework == "onnx":
+            self._framework = "onnx"
+            model_loc = CrossEncoderModel._trace_onnx(model, features, mname)
+        else:
+            raise Exception(
+                f"Unrecognized framework {framework}. Accepted values are `pt`, `onnx`"
+            )
+
+        # save tokenizer file
+        tk_path = Path(f"/tmp/{mname}-tokenizer")
+        tk.save_pretrained(tk_path)
+        if tk.model_max_length > model.get_max_length():
+            model_config = AutoConfig.from_pretrained(self._hf_model_id)
+            if hasattr(model_config, "max_position_embeddings"):
+                tk.model_max_length = model_config.max_position_embeddings
+            elif hasattr(model_config, "n_positions"):
+                tk.model_max_length = model_config.n_positions
+            else:
+                tk.model_max_length = 2**15  # =32768. Set to something big I guess
```

Comment:
Setting an arbitrary value doesn't seem like a good solution.

What do you think about following this: https://github.com/opensearch-project/opensearch-py-ml/blob/main/opensearch_py_ml/ml_models/sentencetransformermodel.py#L936-L942

Reply:
would love to. Unfortunately, `model.get_max_seq_length()` is not a method of most hf transformer ModelForSequenceClassification - that's a special thing from the sentence-transformers model interface, and I'm not sure it is the guarantee that it claims to be: [[implementation]](https://github.com/UKPLab/sentence-transformers/blob/1eec036cdcce649502de6152cb8a3b1ea1bb9a8c/sentence_transformers/SentenceTransformer.py#L436)

Reply:
I'm seeing two problems here:

1. tk.model_max_length is None:    https://github.com/huggingface/transformers/issues/14561 ; if you see this, `model_max_length` will be a larger value and we are going to set a large value which model can't process

2. I'm not a big fan of setting a static big value like `tk.model_max_length = 2**15`. What exactly are we achieving here?

It's important to align `model_max_length` with the model's actual capability (`max_position_embeddings`). Setting `model_max_length` to a higher value than `max_position_embeddings` could lead to errors or unexpected behavior, as the model won't be able to handle inputs longer than its architectural limit. Conversely, setting a lower `model_max_length` can be useful for optimizing performance or adhering to specific task requirements.

Reply:
I don't understand these issues. `tk.model_max_length is None` is exactly the condition that triggers setting it to something reasonable. I agree that it's important to set it to a value that agrees with the model's capability: that's why the first condition here is checking and setting to `max_position_embeddings`. We only pick the very large number when we don't have a bound. For instance, suppose someone trained a mamba-based cross encoder with infinite context cuz it's a RNN.

Reply:
Interesting. Can we just let huggingface/transformers fix this bug? It seems like a them problem, and from what I can tell the only times we're gonna hit it is if someone is trying to use a very old tokenizer file with their thing. At that point I hope we can assume the user is proficient enough with transformers to debug if necessary.